pkg/privilege/privileges/privileges.go

```diff
@@ -207,14 +212,77 @@ func (p *UserPrivileges) RequestVerificationWithUser(db, table, column string, p
+func (p *UserPrivileges) checkMySQLAndPluginDynamicPrivilege(mysqlPriv *MySQLPrivilege, activeRoles []*auth.RoleIdentity, privName string, grantable bool) bool {
+	mysqlChecker := func() bool {
+		return mysqlPriv.RequestDynamicVerification(activeRoles, p.user, p.host, privName, grantable)
+	}
+	pluginChecker := func(authPlugin *extension.AuthPlugin) bool {
+		if authPlugin.VerifyDynamicPrivilege == nil {
+			return true
+		}
+		return authPlugin.VerifyDynamicPrivilege(activeRoles, p.user, p.host, privName, grantable, p.sessionVars.TLSConnectionState)
+	}
+	return p.checkMySQLAndPluginPrivilegeInternal(mysqlChecker, pluginChecker)
+}
+
+// checkMySQLAndPluginPrivilege checks privilege for MySQL, and then checks the extension auth plugin.
+func (p *UserPrivileges) checkMySQLAndPluginPrivilege(mysqlPriv *MySQLPrivilege, user, host string, activeRoles []*auth.RoleIdentity, db, table, column string, priv mysql.PrivilegeType) bool {
+	mysqlChecker := func() bool {
+		return mysqlPriv.RequestVerification(activeRoles, user, host, db, table, column, priv)
+	}
+	pluginChecker := func(authPlugin *extension.AuthPlugin) bool {
+		if authPlugin.VerifyPrivilege == nil {
+			return true
+		}
+		return authPlugin.VerifyPrivilege(activeRoles, user, host, db, table, column, priv, p.sessionVars.TLSConnectionState)
+	}
+	return p.checkMySQLAndPluginPrivilegeInternal(mysqlChecker, pluginChecker)
+}
+
+// checkMySQLAndPluginPrivilegeInternal checks privilege for MySQL, and then checks the extension auth plugin by using the provided checkers.
+func (p *UserPrivileges) checkMySQLAndPluginPrivilegeInternal(mysqlPrivilegeChecker func() bool, pluginPrivilegeChecker func(*extension.AuthPlugin) bool) bool {
+	mysqlAllowed := mysqlPrivilegeChecker()
+	// If userRecord or sessionVars is nil, it means the user is not authenticated using extension auth plugins. Skip the extension auth plugin check.
+	if p.userRecord == nil || p.sessionVars == nil {
+		return mysqlAllowed
+	}
+	// If MySQL privilege check failed, skip the extension auth plugin check.
+	if !mysqlAllowed {
+		return false
+	}
+
+	// Now we know that user has the privilege in MySQL, we need to check the extension auth plugin.
+	if authPlugin, ok := p.authPlugins[p.userRecord.AuthPlugin]; ok {
+		return pluginPrivilegeChecker(authPlugin)
+	}
+	// This normally should not happen since users can only get to here if they have the privilege in MySQL and are using a registered auth plugin.
+	return mysqlAllowed
+}
+
+func (p *UserPrivileges) authenticateWithPlugin(user *auth.UserIdentity, authUser string, authentication, salt []byte, sessionVars *variable.SessionVars, authConn conn.AuthConn, authPlugin *extension.AuthPlugin, pwd string, record *UserRecord, hasPassword string) error {
+	if err := authPlugin.AuthenticateUser(authUser, pwd, authentication, salt, sessionVars.TLSConnectionState, authConn); err != nil {
```

Comment:
I'm not sure we should both pass `user/host` and `authUser/authHost` to method `AuthenticateUser`. They are not all the same.

Reply:
Removed the `authUser` parameter. That one seems to be only used for LDAP based authentications.